datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -557,9 +566,44 @@ private void getAssignmentsFuture(List<Future<Boolean>> assignmentChangeFutures,
   public void onAssignmentChange() {
     _log.info("Coordinator::onAssignmentChange is called");
     _eventQueue.put(CoordinatorEvent.createHandleAssignmentChangeEvent());
+    _assignmentRetryCount.set(0);
     _log.info("Coordinator::onAssignmentChange completed successfully");
   }
 
+  private int getAssignmentTaskCount(Map<String, List<DatastreamTask>> assignment) {
+    return assignment.values().stream().mapToInt(List::size).sum();
+  }
+
+  private void retryHandleAssignmentChange(boolean isDatastreamUpdate) {
+    boolean queueEvent = true;
+    if (_assignmentRetryCount.compareAndSet(ASSIGNMENT_MAX_RETRY_COUNT, 0)) {
+      _log.error(
+          "onAssignmentChange retries reached threshold of {}. Not queuing further retry on onAssignmentChange event.",
+          ASSIGNMENT_MAX_RETRY_COUNT);
+      queueEvent = false;
+    }
+
+    if (isDatastreamUpdate) {
+      _log.warn("Updating metric for event " + CoordinatorEvent.EventType.HANDLE_DATASTREAM_CHANGE_WITH_UPDATE);
+      _metrics.updateKeyedMeter(CoordinatorMetrics.getKeyedMeter(
+          CoordinatorEvent.EventType.HANDLE_DATASTREAM_CHANGE_WITH_UPDATE), 1);
+      if (queueEvent) {
+        _log.warn("Queuing onAssignmentChange event");
+        _eventQueue.put(CoordinatorEvent.createHandleDatastreamChangeEvent());
+        _assignmentRetryCount.incrementAndGet();
+      }
+    } else {
+      _log.warn("Updating metric for event " + CoordinatorEvent.EventType.HANDLE_ASSIGNMENT_CHANGE);
+      _metrics.updateKeyedMeter(CoordinatorMetrics.getKeyedMeter(
+          CoordinatorEvent.EventType.HANDLE_ASSIGNMENT_CHANGE), 1);
+      if (queueEvent) {
+        _log.warn("Queuing onAssignmentChange event");
+        _eventQueue.put(CoordinatorEvent.createHandleAssignmentChangeEvent());
+        _assignmentRetryCount.incrementAndGet();
+      }
+    }
```

Comment:
Based on `isDatastreamUpdate` you can create 2 variables eventType and Event and directly use them and avoid the duplicate code.